src/Graph.ts

```diff
@@ -106,15 +104,44 @@ export default class Graph {
 			this.isPureExternalModule = () => false;
 		}
 
+		this.contextParse = (code: string, options: acorn.Options = {}) => {
+			return this.acornParse(
+				code,
+				Object.assign({}, defaultAcornOptions, options, this.acornOptions)
+			);
+		};
+
+		this.pluginContext = {
+			resolveId: undefined,
+			parse: this.contextParse,
+			warn(warning: RollupWarning | string) {
+				if (typeof warning === 'string') warning = { message: warning };
+				this.warn(warning);
+			},
+			error(err: RollupError | string) {
+				if (typeof err === 'string') throw new Error(err);
+				error(err);
+			}
+		};
+
 		this.resolveId = first(
 			[
 				((id: string, parentId: string) =>
-					this.isExternal(id, parentId, false) ? false : null) as ResolveIdHook
+					this.isExternal.call(this.pluginContext, id, parentId, false)
+						? false
+						: null) as ResolveIdHook
 			]
-				.concat(this.plugins.map(plugin => plugin.resolveId).filter(Boolean))
+				.concat(
+					this.plugins
+						.map(plugin => plugin.resolveId)
+						.filter(Boolean)
+						.map(resolveId => resolveId.bind(this.pluginContext))
+				)
 				.concat(resolveId(options))
 		);
 
+		this.pluginContext.resolveId = this.resolveId;
+
 		const loaders = this.plugins.map(plugin => plugin.load).filter(Boolean);
```

Comment:
I would prefer adding the context to the loaders here instead of having to remember this everywhere where `this.load` is called, or `graph.load` for that matter.

Reply:
With regards to this feedback on having the context bound - the issue here is that we then have to duplicate the types of everything as one type will have the 'this' binding and one type without. Personally I think the type maintenance is more work than just applying the right call, and TypeScript will error if the right call context isn't provided as well.